### ISSUE TITLE
[Question][Help Wanted] Actor and critic extractor outputs lose leaf status when iterating over batch input

### ISSUE THREAD

**Issue by JustinS6626** [DLR-RM/stable-baselines3#1560]
### ❓ Question

I am trying to do an experiment with a PPO algorithm that uses a custom architecture for both the actor and critic, and because of the nature of the algorithm that I am trying to test, I am not able to use a network that will simultaneously accommodate the shapes of the buffer roll-out inputs and the batch inputs sampled from the buffer simultaneously. For this reason, I have to iterate the actor and critic networks over the indices of the batch input, and stack the outputs into a single tensor for use in calculating loss. Here is a toy example for how I am doing that:

```
from typing import Callable, Dict, List, Optional, Tuple, Type, Union

from gymnasium import spaces
import torch as th
from torch import nn

from stable_baselines3 import PPO
from stable_baselines3.common.policies import ActorCriticPolicy

class CustomNetwork(nn.Module):
    def __init__(
        self,
        feature_dim: int,
        last_layer_dim_pi: int = 64,
        last_layer_dim_vf: int = 64,
    ):
        super().__init__()
        self.feature_dim = feature_dim
        self.latent_dim_pi = last_layer_dim_pi
        self.latent_dim_vf = last_layer_dim_vf
        self.policy_net = nn.Sequential(
            nn.Linear(feature_dim, last_layer_dim_pi), nn.ReLU()
        )
        self.value_net = nn.Sequential(
            nn.Linear(feature_dim, last_layer_dim_vf), nn.ReLU()
        )
    def forward(self, features: th.Tensor) -> Tuple[th.Tensor, th.Tensor]:
        x_in = features.view(features.shape[0], 1, features.shape[1])
        return self.forward_actor(x_in), self.forward_critic(x_in)

    def forward_actor(self, features: th.Tensor) -> th.Tensor:
        out_list = []
        for i in range(features.shape[0]):
            feature_input = th.index_select(features, 0, th.tensor(i).cuda())
            result = self.policy_net(feature_input)
            out_list.append(result)
        out = th.stack(out_list)

        print("Actor output is leaf: " + str(out.is_leaf))
        return out

    def forward_critic(self, features: th.Tensor) -> th.Tensor:
        out_list = []
        for i in range(features.shape[0]):
            feature_input = th.index_select(features, 0, th.tensor(i).cuda())
            result = self.value_net(feature_input)
            out_list.append(result)
        out = th.stack(out_list)
        print("Critic output is leaf: " + str(out.is_leaf))
        return self.value_net(features)

class CustomActorCriticPolicy(ActorCriticPolicy):
    def __init__(
        self,
        observation_space: spaces.Space,
        action_space: spaces.Space,
        lr_schedule: Callable[[float], float],
        *args,
        **kwargs,
    ):

        super().__init__(
            observation_space,
            action_space,
            lr_schedule,
            *args,
            **kwargs,
        )
        self.ortho_init = False

    def _build_mlp_extractor(self) -> None:
        self.mlp_extractor = CustomNetwork(self.features_dim)


model = PPO(CustomActorCriticPolicy,"CartPole-v1" , verbose=1)
model.learn(5000)
```
The `print` commands in the `forward_actor` and `forward_critic` methods print out a message indicating whether the tensor being returned is a leaf and will generate a gradient when used in a loss function. For the rollout inputs, I get the messages `Actor output is leaf: True` and `Critic output is leaf: True`. However, during the training phase when the custom network receives the batch inputs, both of the messages show false, which, I think, means that the parameters of the model will not be updated because there is no gradient. Based on this, I am wondering what is different about the batch inputs, apart from their size, and how I can get around this so that the outputs from iterating over them will generate a gradient?

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Closing as we do not provide tech support (see issue template and readme) and as you posted your question at the correct place too (RL discord).

**Comment by JustinS6626** [comment]
I understand. Apologies for the mistake, but I really need some help on this, as it's a major hurdle for the experiment I'm working on now. If you would be willing to provide a reply to my post on the RL discord, I would really appreciate it.
